### DISCUSSION TITLE
How to get all tables/models from Select object?

### DISCUSSION THREAD

**Discussion by slymit** [sqlalchemy/sqlalchemy#8078] [Usage Questions] [answered]
Hello! 
How to get all tables/models from Select object?
I am using a module visitors to find all models from an Select object (based on [this solution](https://stackoverflow.com/a/66855484)):
```python
for visitor in visitors.iterate(stmt):
    if visitor.__visit_name__ == 'binary':
        for v in visitors.iterate(visitor):
            try:
                model = get_model_from_table(v.table)
            except AttributeError:
                pass
            else:
                if model not in models:
                    models.append(model)

    if visitor.__visit_name__ == 'table':
        model = visitor.entity_namespace
        if isinstance(model, DeclarativeMeta) and model not in models:
            models.append(model)

    if visitor.__visit_name__ == 'column':
        try:
            model = get_model_from_table(visitor.table)
        except AttributeError:
            pass
        else:
            if model not in models:
                models.append(model)
```
But I don't think it's a perfect solution.

**Comment by CaselIT** [comment]
Do there not work for you?
- https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=froms#sqlalchemy.sql.expression.Select.columns_clause_froms
- https://docs.sqlalchemy.org/en/14/core/selectable.html?highlight=froms#sqlalchemy.sql.expression.Select.get_final_froms

**Comment by zzzeek** [comment, marked as answer]
note also that to get "models", the usual API to use is `column_descriptions`, see the tutorial at https://docs.sqlalchemy.org/en/14/orm/queryguide.html#queryguide-inspection 


that will give you the entities being **selected from** to be returned in the result set.  if you're trying to find models that are buried elsewhere in the statement, like in joins and such, it would be better if you could be more specific what it is you're trying to do.

**Reply by slymit** [reply]
Thanks for the answer.
I want find all models that are participant in the statement (froms, select_from, joins and such).

**Reply by zzzeek** [reply]
I would take the implementation from https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/util.py#L326 which will find everything for you.  however, everything is returned in terms of Core Table, Join, etc.   the ORM models are within the _annotations collections of Table and other objects, which you can take a look at, but this is not public API right now.    (we can add public API to get the entity from a Core fromclause, we just don't have that right now).

**Reply by slymit** [reply]
Thanks! This is what I need )
